### DISCUSSION TITLE
x509 certificate validator problem

### DISCUSSION THREAD

**Discussion by gfhoffmann** [ipxe/ipxe#1087] [Q&A]
Hey there!

I'm having this weird problem with my iPXE solution. I'm trying to chain a public URL (its a public bucket on Cloudflare) that will proceed with the installation steps. The problem is that I have machines in different Subnets and IP Prefixes.

I'm noticing that machines from one specific IP Prefix are not being able to validate the certificate of the public URL that we are trying to chain. But for other machines the iPXE is able to validate and proceeds normally.

Here I have the screenshot of the failing machine (I have a retry on the chain, so that's why its showing duplicated in some areas, but the machine stays in this state not being able to validate forever):

![image](https://github.com/ipxe/ipxe/assets/105290508/72e1ef9b-dcb1-47fa-b023-e1322dbe5364)

And here I have two screenshots of the working machine. As we can see, it failed sometimes to validate but then it worked.

![image](https://github.com/ipxe/ipxe/assets/105290508/ddc1290b-3d52-4a6b-a82e-dc9dfd57dadb)

![image](https://github.com/ipxe/ipxe/assets/105290508/880a6827-68cd-42e0-b465-89c5ba7ba7f6)

What can I do to discover why this is happening and how can I fix it?

Thanks!

**Comment by mcb30** [comment]
As a first step, go to the error page https://ipxe.org/420c6001 that iPXE is telling you to look at.

That page suggests building with `DEBUG=httpcore`, which may give you more information.

**Reply by gfhoffmann** [reply]
Hello Michael! Thanks for answering.

In the screenshots that I posted the iPXE already has `DEBUG=tls,x509,validator,httpcore` on.

The error page doesn't tell how to resolve this (or give any tips/clues). Just a line of code that raised the error, which I can do nothing about it.

Any ideas in how to proceed?

**Reply by mcb30** [reply]
OK.  In that case, it looks as though the TCP connection is being closed unexpectedly by the server.

You could build with `DEBUG=tcp` or `DEBUG=tcp:3`.  This should give some (verbose) information about what iPXE thinks is happening at the TCP layer.  It's likely to be quite messy output, since there will be several open connections at the point of interest, so you may want to try to capture as serial console output, rather than as screenshots.

**Reply by gfhoffmann** [reply]
Perfect! I can do that! I will watch the execution via serial console instead of the browser.

I will be back with more info, thanks!

**Reply by gfhoffmann** [reply]
I'm back! I did my best to get the logs, let's see if it helps.

```
Intel(R) Boot Agent XE v2.3.98
Copyright (C) 1997-2017, Intel Corporation

PXELINUX 6.04 lwIP 20191223 Copyright (C) 1994-2015 H. Peter Anvin et al
Booting local disk ...
Booting...
iPXE initialising devices...ok

iPXE 1.21.1+ (gff0f8) -- Open Source Network Boot Firmware -- https://ipxe.org
Features: DNS HTTP HTTPS iSCSI NFS TFTP SRP AoE ELF MBOOT PXE bzImage Menu PXEXT

Getting the script url
TLS 0x104b54 created session bucket.public.url.here
https://bucket.public.url.here/machine__name.ipxe.................. Connection timed out
(https://ipxe.org/4c116035)
Getting the script url
TLS 0x103b54 created session bucket.public.url.here
https://bucket.public.url.here/machine__name.ipxe................. [bucket.public.url.here]TCP
0x0 RX 60217<-38250           00000000 8622896b..8622896c    0 SYN
. Connection timed out (https://ipxe.org/4c116035)
X509 chain 0x1001a4 added X509 0x104004 "bucket.public.url.here"
TLS 0x100b64 found certificate bucket.public.url.here
X509 chain 0x1001a4 added X509 0x105004 "Cloudflare Inc RSA CA-2"
TLS 0x100b64 found certificate Cloudflare Inc RSA CA-2
TCP 0x1019c4 TX 18612->443 35dd57af..35dd57af           ad69961c    0 ACK
TCP 0x1019c4 RX 18612<-443           35dd57af ad699018..ad6995a8 1424 ACK
TCP 0x1019c4 RX 18612<-443           35dd57af ad698a88..ad699018 1424 ACK
TLS 0x100b64 using protocol version 3.3
TLS 0x100b64 selected pubkey-rsa-aes_gcm-128-sha256
TCP 0x1019c4 RX 18612<-443           35dd57af ad6995a8..ad69961c  116 PSH ACK
TLS 0x100b64 found certificate Cloudflare Inc RSA CA-2
TCP 0x1019c4 TX 18612->443 35dd57af..35dd57af           ad69961c    0 ACK
TCP 0x1019c4 RX 18612<-443           35dd57af ad699018..ad6995a8 1424 ACK
X509 chain 0x1001a4 found no usable certificates
TCP 0x1019c4 TX 18612->443 35dd57af..35dd57af           ad69961c    0 ACK
TCP 0x1013f4 allocated
TCP 0x1013f4 transitioned from CLOSED to SYN_SENT
TCP 0x1013f4 bound to port 62650
TCP 0x1013f4 using timestamps, SACK, TX window x128, RX window x512
TCP 0x1013f4 transitioned from SYN_SENT to ESTABLISHED
TCP 0x1013f4 TX 62650->80 63363cc3..63363db9           59b040d0  246 PSH ACK
TCP 0x1013f4 RX 62650<-80           63363db9 59b040d0..59b040d0    0 ACK
TCP 0x1013f4 RX 62650<-80           63363db9 59b04c20..59b04d36  278 PSH ACK
TCP 0x1013f4 TX 62650->80 63363db9..63363db9           59b040d0    0 ACK
X509 chain 0x1001a4 added X509 0x104004 "bucket.public.url.here"
TLS 0x100b64 found certificate bucket.public.url.here
X509 chain 0x1001a4 added X509 0x105004 "Cloudflare Inc RSA CA-2"
TLS 0x100b64 found certificate Cloudflare Inc RSA CA-2
TCP 0x1019c4 TX 18612->443 35dd57af..35dd57af           ad69961c    0 ACK
TCP 0x1019c4 RX 18612<-443           35dd57af ad699018..ad6995a8 1424 ACK
X509 chain 0x1001a4 found no usable certificates
VALIDATOR 0x100df4 "bucket.public.url.here" downloading "Cloudflare Inc RSA CA-2" cros
s-signature from http://ca.ipxe.org/auto/34a7304c.der?subject=MFoxCzAJBgNVBAYTAk
lFMRIwEAYDVQQKEwlCYWx0aW1vcmUxEzARBgNVBAsTCkN5YmVyVHJ1c3QxIjAgBgNVBAMTGUJhbHRpbW
9yZSBDeWJlclRydXN0IFJvb3Q=
TCP 0x1019c4 TX 18612->443 35dd57af..35dd57af           ad69961c    0 ACK
TCP 0x1013f4 allocated
TCP 0x1013f4 transitioned from CLOSED to SYN_SENT
TCP 0x1013f4 bound to port 62650
TCP 0x1013f4 timer fired in SYN_SENT for 63363cc2..63363cc2 00000000
TCP 0x1013f4 TX 62650->80 63363cc2..63363cc3           00000000    0 SYN
TCP 0x1013f4 RX 62650<-80           63363cc3 59b040cf..59b040d0    0 SYN ACK
TCP 0x1013f4 using timestamps, SACK, TX window x128, RX window x512
TCP 0x1013f4 transitioned from SYN_SENT to ESTABLISHED
TCP 0x1013f4 TX 62650->80 63363cc3..63363db9           59b040d0  246 PSH ACK
TCP 0x1013f4 RX 62650<-80           63363db9 59b040d0..59b040d0    0 ACK
TCP 0x1013f4 RX 62650<-80           63363db9 59b04c20..59b04d36  278 PSH ACK
TCP 0x1013f4 TX 62650->80 63363db9..63363db9           59b040d0    0 ACK
..... [XCRT Cloudflare Inc RSA CA-2]TCP 0x1013f4 RX 62650<-80           63363db9
 59b04d36..59b04d37    0 FIN ACK
TCP 0x1013f4 TX 62650->80 63363db9..63363db9           59b040d0    0 ACK
... [XCRT Cloudflare Inc RSA CA-2]TCP 0x0 RX 12352<-56752           00000000 d5d
623b2..d5d623b3    0 SYN
...... [XCRT Cloudflare Inc RSA CA-2]TCP 0x1019c4 RX 18612<-443           35dd57
af ad69961c..ad69961d    0 FIN ACK
TCP 0x1013f4 transitioned from ESTABLISHED to FIN_WAIT_1
TCP 0x1013f4 TX 62650->80 63363db9..63363dba           59b040d0    0 FIN ACK
TCP 0x1019c4 transitioned from ESTABLISHED to CLOSING/LAST_ACK
 Error 0x420c6001 (https://ipxe.org/420c6001)
TCP 0x1019c4 TX 18612->443 35dd57af..35dd57b0           ad69961d    0 FIN ACK
...... [XCRT Cloudflare Inc RSA CA-2]TCP 0x1019c4 RX 18612<-443           35dd57
TCP 0x1013f4 TX 62650->80 63363db9..63363dba           59b040d0    0 FIN ACK
TCP 0x1019c4 transitioned from ESTABLISHED to CLOSING/LAST_ACK
TCP 0x1013f4 timer fired in FIN_WAIT_1 for 63363db9..63363dba 59b040d0
TCP 0x1013f4 TX 62650->80 63363db9..63363dba           59b040d0    0 FIN ACK
TCP 0x1013f4 RX 62650<-80           63363dba 59b04d37..59b04d37    0 ACK
TCP 0x1013f4 transitioned from FIN_WAIT_1 to FIN_WAIT_2
TCP 0x1013f4 TX 62650->80 63363dba..63363dba           59b040d0    0 ACK
TCP 0x1019c4 RX 18612<-443           35dd57af ad69961c..ad69961d    0 FIN ACK
TCP 0x1019c4 timer fired in CLOSING/LAST_ACK for 35dd57af..35dd57b0 ad69961d
TCP 0x1019c4 transitioned from CLOSING/LAST_ACK to TIME_WAIT
TCP 0x1013f4 RX 62650<-80           63363dba 59b04d37..59b04d37    0 ACK
TCP 0x1013f4 TX 62650->80 63363dba..63363dba           59b040d0    0 ACK
X509 chain 0x1006a4 added X509 0x104004 "bucket.public.url.here"
TLS 0x100c64 found certificate bucket.public.url.here
X509 chain 0x1006a4 added X509 0x105004 "Cloudflare Inc RSA CA-2"
TLS 0x100c64 found certificate Cloudflare Inc RSA CA-2
TCP 0x101234 TX 56772->443 1c91107d..1c91107d           8dfa2f5d    0 ACK
TCP 0x101234 RX 56772<-443           1c91107d 8dfa2ee9..8dfa2f5d  116 PSH ACK
X509 chain 0x1006a4 found no usable certificates
VALIDATOR 0x100ef4 "bucket.public.url.here" downloading "Cloudflare Inc RSA CA-2" cros
s-signature from http://ca.ipxe.org/auto/34a7304c.der?subject=MFoxCzAJBgNVBAYTAk
lFMRIwEAYDVQQKEwlCYWx0aW1vcmUxEzARBgNVBAsTCkN5YmVyVHJ1c3QxIjAgBgNVBAMTGUJhbHRpbW
9yZSBDeWJlclRydXN0IFJvb3Q=
TCP 0x101234 TX 56772->443 1c91107d..1c91107d           8dfa2f5d    0 ACK
TCP 0x108804 allocated
TCP 0x108804 transitioned from CLOSED to SYN_SENT
TCP 0x108804 bound to port 62270
TCP 0x108804 timer fired in SYN_SENT for 02234c1f..02234c1f 00000000
TCP 0x108804 TX 62270->80 02234c1f..02234c20           00000000    0 SYN
TCP 0x108804 RX 62270<-80           02234c20 a84d60b5..a84d60b6    0 SYN ACK
TCP 0x108804 using timestamps, SACK, TX window x128, RX window x512
TCP 0x108804 transitioned from SYN_SENT to ESTABLISHED
TCP 0x108804 TX 62270->80 02234c20..02234d16           a84d60b6  246 PSH ACK
TCP 0x108804 RX 62270<-80           02234d16 a84d60b6..a84d60b6    0 ACK
TCP 0x108804 RX 62270<-80           02234d16 a84d6c06..a84d6d1c  278 PSH ACK
TCP 0x108804 TX 62270->80 02234d16..02234d16           a84d60b6    0 ACK
..... [XCRT Cloudflare Inc RSA CA-2]TCP 0x108804 RX 62270<-80           02234d16
 a84d6d1c..a84d6d1d    0 FIN ACK
TCP 0x108804 TX 62270->80 02234d16..02234d16           a84d60b6    0 ACK
TCP 0x0 RX 56864<-56125           00000000 99000e1b..99000e1c    0 SYN
TCP 0x0 RX 22<-61000           00000000 0daddeec..0daddeed    0 SYN
... [XCRT Cloudflare Inc RSA CA-2]
```

I didn't paste all the logs because it kept repeating itself.
Hope it helps.

**Reply by NiKiZe** [reply]
Drop other debug options, what is the issue without any debug options?
Does this machine have full internet access, or is it limited in any way?

**Reply by gfhoffmann** [reply]
Ok, I will run again just with httpcore and tcp:3

The machine has full access to internet, using a "normal" Ubuntu installation we can access the public bucket, etc.

The issue without any debug is just the https://ipxe.org/420c6001 error.

**Comment by mcb30** [comment]
> I'm back! I did my best to get the logs, let's see if it helps.

Thank you!  Dissecting the logs:

> ```
> VALIDATOR 0x100df4 "bucket.public.url.here" downloading "Cloudflare Inc RSA CA-2" cross-signature from http://ca.ipxe.org/auto/34a7304c.der?subject=MFoxCzAJBgNVBAYTAklFMRIwEAYDVQQKEwlCYWx0aW1vcmUxEzARBgNVBAsTCkN5YmVyVHJ1c3QxIjAgBgNVBAMTGUJhbHRpbW9yZSBDeWJlclRydXN0IFJvb3Q=
> ```

This is iPXE deciding that it doesn't have a complete certificate chain, and starting the attempt to download the relevant cross-signing certificate to complete the chain.

> ```
> TCP 0x1013f4 allocated
> TCP 0x1013f4 transitioned from CLOSED to SYN_SENT
> TCP 0x1013f4 bound to port 62650
> TCP 0x1013f4 timer fired in SYN_SENT for 63363cc2..63363cc2 00000000
> TCP 0x1013f4 TX 62650->80 63363cc2..63363cc3           00000000    0 SYN
> TCP 0x1013f4 RX 62650<-80           63363cc3 59b040cf..59b040d0    0 SYN ACK
> TCP 0x1013f4 using timestamps, SACK, TX window x128, RX window x512
> TCP 0x1013f4 transitioned from SYN_SENT to ESTABLISHED
> TCP 0x1013f4 TX 62650->80 63363cc3..63363db9           59b040d0  246 PSH ACK
> TCP 0x1013f4 RX 62650<-80           63363db9 59b040d0..59b040d0    0 ACK
> ```

This is the TCP-level debug for the HTTP connection used to download the cross-signed certificate.  So far we see the SYN and SYN+ACK handshake, and the request itself (in the PSH+ACK packet).

> ```
> TCP 0x1013f4 RX 62650<-80           63363db9 59b04c20..59b04d36  278 PSH ACK
> TCP 0x1013f4 TX 62650->80 63363db9..63363db9           59b040d0    0 ACK
> TCP 0x1013f4 RX 62650<-80           63363db9 59b04d36..59b04d37    0 FIN ACK
> TCP 0x1013f4 TX 62650->80 63363db9..63363db9           59b040d0    0 ACK
> ```

Here we see the response from the server, and the connection being closed by the server (in the FIN+ACK).  Interestingly, the total response size is only 278 bytes.  This is far too small to contain the cross-signed certificate (and its HTTP headers).

> ```
> TCP 0x0 RX 12352<-56752           00000000 d5d623b2..d5d623b3    0 SYN
> ```

This is probably irrelevant, but we do at this point see something attempting to initiate a TCP connection to iPXE on port 12352.  No idea what this could be.  iPXE does not support listening sockets, so this is not something iPXE has asked for.  Could be some kind of monitoring system that you have set up in your subnet?

> ```
> TCP 0x1019c4 RX 18612<-443           35dd57af ad69961c..ad69961d    0 FIN ACK
> TCP 0x1019c4 transitioned from ESTABLISHED to CLOSING/LAST_ACK
>  Error 0x420c6001 (https://ipxe.org/420c6001)
> ```

Here we see the HTTPS server closing its TCP connection (in the FIN+ACK packet).  It's worth noting that there has been no data sent or received on this TCP connection since iPXE started the attempt to obtain the cross-signed certificate.  This is the server unilaterally deciding to close the TCP connection, which iPXE then correctly reports as error https://ipxe.org/420c6001 (which, as that page explains, just means that a TCP connection has been closed).

There are no timestamps in the logs, so we're not able to tell if this is the server misbehaving, or the server deciding that it's waited too long for a response from iPXE.


There's definitely something strange about the download of the cross-signed certificate, since the response is too small to be correct.  I'd probably try next with `DEBUG=validator,x509,tls,httpcore:3` - that should show you the transmitted and received HTTP headers in all connections (including the cross-signed certificate download attempt).

**Reply by gfhoffmann** [reply]
Wow! Thanks for your response @mcb30!

I will run more tests with the DEBUG that you mentioned.

I'm having a hard time to save the logs from SOL (Serial Over Lan). The iPXE refreshes the screen all the time, and when I open the log file (using `cat`), it only shows the last thing that was appearing in the screen.

I was able to open the file in VSCode, but it was full of strange chars.

Do you have any tips in how can I save this in a different way to help you guys with a cleaner log?

Thanks!

**Reply by gfhoffmann** [reply]
@mcb30 I was able to get HTTP infos using `httpcore:3`

The only thing related to HTTP was this:

```
HTTP 0x101d54 TX GET /auto/34a7304c.der?subject=MFoxCzAJBgNVBAYTAklFMRIwEAYDVQQK
EwlCYWx0aW1vcmUxEzARBgNVBAsTCkN5YmVyVHJ1c3QxIjAgBgNVBAMTGUJhbHRpbW9yZSBDeWJlclRy
dXN0IFJvb3Q= HTTP/1.1
HTTP 0x101d54 TX Connection: keep-alive
HTTP 0x101d54 TX User-Agent: iPXE/1.21.1+ (gff0f8)
HTTP 0x101d54 TX Host: ca.ipxe.org
```

**Reply by mcb30** [reply]
> I'm having a hard time to save the logs from SOL (Serial Over Lan). The iPXE refreshes the screen all the time, and when I open the log file (using `cat`), it only shows the last thing that was appearing in the screen.

iPXE shouldn't be doing anything to refresh or clear the screen (unless you have your script deliberately echoing control characters), so this sounds like an artifact from the serial-over-LAN capture.

> Do you have any tips in how can I save this in a different way to help you guys with a cleaner log?

You may be able to get more visibility using `cat -v` to avoid having any embedded control sequences rendered directly.

> @mcb30 I was able to get HTTP infos using `httpcore:3`
> 
> The only thing related to HTTP was this:
> 
> ```
> HTTP 0x101d54 TX GET /auto/34a7304c.der?subject=MFoxCzAJBgNVBAYTAklFMRIwEAYDVQQK
> EwlCYWx0aW1vcmUxEzARBgNVBAsTCkN5YmVyVHJ1c3QxIjAgBgNVBAMTGUJhbHRpbW9yZSBDeWJlclRy
> dXN0IFJvb3Q= HTTP/1.1
> HTTP 0x101d54 TX Connection: keep-alive
> HTTP 0x101d54 TX User-Agent: iPXE/1.21.1+ (gff0f8)
> HTTP 0x101d54 TX Host: ca.ipxe.org
> ```

This cannot be all that there is, since we already know that there is **some** kind of response to that HTTP request.
